### PR TITLE
Add project directory validation for balena push / build / deploy

### PR DESCRIPTION
**Pull request by pdcastro**
Connects-to: #1421
Resolves: #1619 
Change-type: minor
Signed-off-by: Paulo Castro <paulo@balena.io>

### CODE REVIEW

File: lib/utils/compose_ts.ts [hunk 1]
```diff
@@ -264,63 +318,161 @@ async function performResolution(
...
+ */
+export async function validateProjectDirectory(
+	sdk: BalenaSDK,
+	opts: {
+		dockerfilePath?: string;
+		noParentCheck: boolean;
+		projectPath: string;
+		registrySecretsPath?: string;
+	},
+): Promise<ProjectValidationResult> {
```
Comment (pdcastro) [inline lib/utils/compose_ts.ts]: The `validateProjectDirectory()` function is the main feature of this PR, and reason for the 'minor' semver.

**Comment by pdcastro** [comment]
@balena-ci retest

**Comment by pdcastro** [comment]
@Page-, just FYI, in case you were planning to convert `build.coffee` or `deploy.coffee` to Typescript, it might be better to wait for this PR to get merged, as it touches those files.

File: tests/test-data/projects/docker-compose/basic/service2/Dockerfile-alt [hunk 2]
```diff
@@ -0,0 +1 @@
+alternative Dockerfile (basic/service2)
```
Comment (srlowe) [inline tests/test-data/projects/docker-compose/basic/service2/Dockerfile-alt]: Does it matter that there's no EOL here?

File: lib/actions/push.ts [hunk 3]
```diff
@@ -367,14 +364,12 @@ export const push: CommandDefinition<
...
-					.nodeify(done);
+				).catch(BuildError, e => {
+					throw new ExpectedError(e.toString());
+				});
 				break;
 			default:
-				exitWithExpectedError(
+				throw new ExpectedError(
 					stripIndent`
 					Build target not recognised. Please provide either an application name or device address.
```
Comment (srlowe) [inline lib/actions/push.ts]: I know this hasn't been changed. But it we are using US English then this should be spelt `recognized`.

File: tests/test-data/projects/docker-compose/basic/service2/Dockerfile-alt (hunk 2 above)
Comment (pdcastro) [reply tests/test-data/projects/docker-compose/basic/service2/Dockerfile-alt]: I will add a line break, 👍  but this Dockerfile is invalid, of course.

File: lib/actions/push.ts (hunk 3 above)
Comment (pdcastro) [reply lib/actions/push.ts]: I will Americani**S**e it :-)

**Pull request merged**